pkg/bindinfo/binding_auto.go

```diff
@@ -0,0 +1,150 @@
+// Copyright 2025 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bindinfo
+
+import (
+	"fmt"
+
+	"github.com/pingcap/tidb/pkg/parser"
+	"github.com/pingcap/tidb/pkg/sessionctx"
+	"github.com/pingcap/tidb/pkg/util"
+	"github.com/pingcap/tidb/pkg/util/chunk"
+	"github.com/pingcap/tidb/pkg/util/logutil"
+	utilparser "github.com/pingcap/tidb/pkg/util/parser"
+	"go.uber.org/zap"
+)
+
+// BindingPlanInfo represents the binding plan info.
+type BindingPlanInfo struct {
+	*Binding
+
+	// Info from StmtStats
+	Plan                 string
+	AvgLatency           float64
+	ExecTimes            int64
+	AvgScanRows          float64
+	AvgReturnedRows      float64
+	LatencyPerReturnRow  float64
+	ScanRowsPerReturnRow float64
+
+	// Recommendation
+	Recommend string
+	Reason    string
+}
+
+// BindingAuto represents a series of APIs that help manage bindings automatically.
+type BindingAuto interface {
+	// TODO: RecordHistPlansAsBindings records the history plans as bindings for qualified queries.
+
+	// ShowPlansForSQL shows historical plans for a specific SQL.
+	ShowPlansForSQL(currentDB, sqlOrDigest, charset, collation string) ([]*BindingPlanInfo, error)
+}
+
+type bindingAuto struct {
+	sPool util.DestroyableSessionPool
+}
+
+func newBindingAuto(sPool util.DestroyableSessionPool) BindingAuto {
+	return &bindingAuto{
+		sPool: sPool,
+	}
+}
+
+// ShowPlansForSQL shows historical plans for a specific SQL.
+func (ba *bindingAuto) ShowPlansForSQL(currentDB, sqlOrDigest, charset, collation string) ([]*BindingPlanInfo, error) {
+	// parse and normalize sqlOrDigest
+	p := parser.New()
+	var normalizedSQL, whereCond string
+	stmtNode, err := p.ParseOneStmt(sqlOrDigest, charset, collation)
+	if err == nil {
+		db := utilparser.GetDefaultDB(stmtNode, currentDB)
+		normalizedSQL, _ = NormalizeStmtForBinding(stmtNode, db, false)
+	}
+
+	// read binding data from mysql.bind_info
+	if normalizedSQL != "" {
+		whereCond = fmt.Sprintf("where original_sql='%s'", normalizedSQL)
```

Comment:
Please use the standard method to escape it since there could be arbitrary content in the SQL.